### PR TITLE
Added getMovie endpoint

### PR DESCRIPTION
**Pull request by jackdevitt**
### Features

- Added endpoint /getMovie (GET)
    - **** How it works
    - Enter a ID or movie Title in the body and it will return data
    - Title returns all movies of similar title (with the closest match at the top)
    - ID returns the exact movie with all the information needed for the game
    - IF BOTH ID and Title are present, it will always use ID and never Title

### Media
Screenshot 2023-07-21 at 1 25 46 PM

### Possible changes
- Make it into two endpoints with one for ID and one for Title

### Known future changes
- Add producer and director to returned info when ID is present
- Prevent movies with fields that cannot work from being returned

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	return entry
+}
+
+func main() {	
+	key = getKey()
+
+	router := gin.Default()
+	router.GET("/getMovie", movieByName)
+	router.Run("0.0.0.0:5050")
+}
```
Comment (matt-j-deasy) [inline main.go]: Right off the bat, I can tell this file is doing too much. When we put all the logic in the one place it becomes tough to follow.  Try to think in terms of having a Model layer which defines the shapes of objects, a controller layer which handles API requests coming in and passing off the logic to a service layer which does the nitty gritty.

So in this example: we may have 
MovieController.go
MovieService.go
etc.

Could talk to James about the normal breakdown here. But I'll hold off on reviewing the rest until we break this file up

**Review by matt-j-deasy — CHANGES_REQUESTED** [review]

File: main.go (hunk 1 above)
Comment (jackdevitt) [reply main.go]: Separated Main.go into three different files:
- controller.go handles the web server and endpoints
- movie.go handles API connections to the movie db
- structs.go handles all structs 

Please let me know if there is something else you would like me to change, thanks!

File: controller.go
Comment (matt-j-deasy) [inline controller.go:1]: Much better! but lets add a qualifier to this file. You wouldn't put all your controllers, structs, etc in 1 large file.  Lets call this movieController.go, and movieStructs.go

**Review by matt-j-deasy — CHANGES_REQUESTED** [review]

File: controller.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+		}
+	}
+}
+
+func main() {	
+	key = getKey()
+
+	router := gin.Default()
+	router.GET("/getMovie", movieByName)
+	router.Run("0.0.0.0:5050")
```
Comment (matt-j-deasy) [inline controller.go]: Where does this port number come from? It seems odd to mention this port without using a variable, either an environment variable or something in Go im not familiar with

File: movie.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+		fmt.Println("ERROR")
+		return ""
+	} else {
+		key = os.Getenv("movieDBKey")
+		return key
+	}
+}
+
+func getMovieByName(title string) MovieDBResponseArray {
+	req := "https://api.themoviedb.org/3/search/movie?api_key=" + key + "&query=" + strings.ReplaceAll(strings.TrimSpace(title), " ", "+")
```
Comment (matt-j-deasy) [inline movie.go]: This url should be a variable. Imagine we end up changing what API we are using to get this info, we would have to change this url everywhere, as opposed to 1 place.

Call this something like: MOVIE_DB_API_ROOT="https://api.themoviedb.org/3/" and then the usage is
"MOVIE_DB_API_ROOT/search/movie?api_key="""

File: movie.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+		fmt.Println("ERROR")
+	} else {
+		body, err := io.ReadAll(response.Body)
+		if err != nil {
+			fmt.Println("ERROR")
+		} else {
+			var entry MovieDetails
+			json.Unmarshal(body, &entry)
+
+			movieActorReq := "https://api.themoviedb.org/3/movie/" + fmt.Sprint(id) + "/credits?api_key=" + key
```
Comment (matt-j-deasy) [inline movie.go]: Same thing about API ROOT

**Review by matt-j-deasy — CHANGES_REQUESTED** [review]

**Review by matt-j-deasy — APPROVED** [review]

**Pull request merged**